File: app/src/main/java/com/ds/ftp/java/ftp/FTPClientUser.java
package com.ds.ftp.java.ftp;

import android.util.Log;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import org.apache.commons.net.ftp.FTP;
import org.apache.commons.net.ftp.FTPClient;

import java.io.IOException;
import java.io.InputStream;
import java.net.SocketException;

/**
 * Class that uses an {@link FTPClient} to check ftp server connection or upload file.
 */
public class FTPClientUser {

    /**
     * Logging tag.
     */
    private final static String TAG = "FTPClientUser";

    /**
     * The [FTPClient].
     */
    private FTPClient ftpClient = new FTPClient();

    /**
     * The server address.
     */
    @Nullable
    private String server;

    /**
     * The server port.
     */
    @Nullable
    private Integer port;

    /**
     * The user's name.
     */
    @Nullable
    private String username;

    /**
     * The password.
     */
    @Nullable
    private String password;

    /**
     * Sets the server address.
     *
     * @param server the server address.
     */
    public void setServer(@Nullable String server) {
        this.server = server;
    }

    /**
     * Sets the server port.
     *
     * @param port the server port.
     */
    public void setPort(@Nullable Integer port) {
        this.port = port;
    }

    /**
     * Sets the user's name.
     *
     * @param username the user's name.
     */
    public void setUsername(@Nullable String username) {
        this.username = username;
    }

    /**
     * Sets the password.
     *
     * @param password the passwrd.
     */
    public void setPassword(@Nullable String password) {
        this.password = password;
    }

    /**
     * Suspend function to check the connection.
     *
     * @return true of connection can be made to the given ftp server, false otherwise.
     */
    public boolean checkConnection() {
        boolean connected = true;
        if (!ftpClient.isConnected()) {
            connected = connect();
        }
        try {
            ftpClient.disconnect();
            return connected;
        } catch (IOException e) {
            return false;
        }
    }

    /**
     * Suspend function to upload given file.
     *
     * @param fileName    the file's name.
     * @param inputStream the input stream pointing at the file.
     * @return true if upload is successful, false otherwise
     */
    public boolean uploadFile(@NonNull final String fileName, @Nullable final InputStream inputStream) {
        if (inputStream == null) {
            return false;
        }
        if (!ftpClient.isConnected()) {
            connect();
        }
        if (!ftpClient.isConnected()) {
            return false;
        } else {
            try {
                boolean success;
                ftpClient.setFileType(FTP.BINARY_FILE_TYPE);
                ftpClient.setFileTransferMode(FTP.STREAM_TRANSFER_MODE);
                ftpClient.enterLocalPassiveMode();
                success = ftpClient.storeFile(fileName, inputStream);
                if (!success) {
                    Log.e(TAG, "reply code: " + ftpClient.getReplyCode() +
                            "  Code message: " + ftpClient.getReplyString());
                }
                ftpClient.disconnect();

                return success;
            } catch (IOException e) {
                return false;
            }
        }
    }

    /**
     * Connects to the given ftp server.
     *
     * @return true if connection was successful, false otherwise.
     */
    private boolean connect() {
        try {
            if (port == null) {
                ftpClient.connect(server);
            } else {
                ftpClient.connect(server, port);
            }
            return ftpClient.login(username, password);
        } catch (SocketException e) {
            return false;
        } catch (IOException e) {
            return false;
        }
    }
}

File: app/src/main/java/com/ds/ftp/kotlin/ftp/FTPClientUser.kt
package com.ds.ftp.kotlin.ftp

import android.util.Log
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.apache.commons.net.ftp.FTP
import org.apache.commons.net.ftp.FTPClient
import java.io.IOException
import java.io.InputStream
import java.net.SocketException

/**
 * Class that uses an [FTPClient] to check ftp server connection or upload file.
 */
class FTPClientUser {

    /**
     * The [FTPClient].
     */
    private val ftpClient: FTPClient = FTPClient()

    /**
     * The server address.
     */
    var server: String? = null

    /**
     * The server port.
     */
    var port: Int? = null

    /**
     * The user's name.
     */
    var username: String? = null

    /**
     * The password.
     */
    var password: String? = null

    /**
     * Suspend function to check the connection.
     *
     * @return true of connection can be made to the given ftp server, false otherwise.
     */
    suspend fun checkConnection(): Boolean {
        return withContext(Dispatchers.Default) {
            val connected = if (!ftpClient.isConnected) {
                connect()
            } else {
                true
            }
            try {
                ftpClient.disconnect()
                connected
            } catch (e: IOException) {
                false
            }
        }
    }

    /**
     * Suspend function to upload given file.
     *
     * @param fileName the file's name.
     * @param inputStream the input stream pointing at the file.
     * @return true if upload is successful, false otherwise
     */
    suspend fun uploadFile(fileName: String, inputStream: InputStream?): Boolean {
        if (inputStream == null) {
            return false
        }
        return withContext(Dispatchers.Default) {
            if (!ftpClient.isConnected) {
                connect()
            }
            if (!ftpClient.isConnected) {
                false
            } else {
                try {
                    val success: Boolean
                    ftpClient.apply {
                        setFileType(FTP.BINARY_FILE_TYPE)
                        setFileTransferMode(FTP.STREAM_TRANSFER_MODE)
                        enterLocalPassiveMode()
                        success = storeFile(fileName, inputStream)
                        if (!success) {
                            Log.e("FTPClientUser", "reply code: $replyCode message: $replyString")
                        }
                        disconnect()
                    }
                    success
                } catch (e: IOException) {
                    false
                }
            }
        }
    }

    /**
     * Connects to the given ftp server.
     *
     * @return true if connection was successful, false otherwise.
     */
    private fun connect(): Boolean {
        ftpClient.apply {
            return try {
                when (port) {
                    null -> connect(server)
                    else -> connect(server, port!!)
                }
                login(username, password)
            } catch (e: SocketException) {
                false
            } catch (e: IOException) {
                false
            }
        }
    }
}